python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,539 @@
+                    queued_tester.setStatus('QUEUED', queued_tester.bucket_pending)
+                    next_job_list.append(next_job_container)
+
+        return next_job_list
+
+    # Queue a job to a specified thread pool
+    def queueJobs(self, status_job_containers=[], tester_job_containers=[]):
+        for job in tester_job_containers:
+            # Assign a thread to do a test
+            self.tester_pool.apply_async(self.testWorker, (job,))
+
+        for job in status_job_containers:
+            # Assign a thread to print some status
+            self.status_pool.apply_async(self.statusWoker, (job,))
+
+    # Status processing (inform TestHarness of a status)
+    def statusWoker(self, job_container):
+        tester = job_container.getTester()
+
+        # Return to the TestHarness
+        self.harness.handleTestStatus(job_container)
+
+        # Decrement the job queue count now that this test has finished
+        if tester.isFinished():
+            with self.thread_lock:
+                self.job_queue_count -= 1
+
+    # Test execution processing
+    def testWorker(self, job_container):
+        tester = job_container.getTester()
+
+        # Check if we have available slots
+        if self.checkAvailableSlots(job_container):
+
+            # Get a timer tuple of started thread timers
+            my_timers = self.handleTimers(start_timers=job_container)
+
+            # Call derived run method outside of the thread_lock (run is blocking)
+            self.run(job_container)
+
+            # The test finished, so stop any timers we received when starting them
+            self.handleTimers(stop_timers=my_timers)
+
+            # Derived run should never set a pending status. If it did, something is wrong as the scheduler can not handle this scenario
+            if tester.isPending():
+                raise SchedulerError('Derived Scheduler can not return a pending status!')
+
+            # Queue this job for status printing
+            self.queueJobs(status_job_containers=[job_container])
+
+            # With this job completed, determine if it failed and if any dependencies will now fail
+            failed_downstream = self.deleteDownstreamTests(job_container)
+
+            # Queue any failing downstream jobs for status printing (skipped dependency)
+            self.queueJobs(status_job_containers=failed_downstream)
+
+            # Get the next group of jobs we can run
+            next_job_group = self.getNextJobGroup(job_container)
+
+            # Recover worker count before attempting to queue more jobs
+            with self.thread_lock:
+                self.workers_in_use = max(0, self.workers_in_use - tester.getProcs(self.options))
+
+            # Queue this new batch of runnable jobs
+            self.queueJobs(tester_job_containers=next_job_group)
+
+        # Not enough slots to run the job
+        else:
```

Comment:
Yeah - this can't be right. One check on the currently running slots and you end up in too big land even if it would fit later.